packages/asset-swapper/contracts/src/DODOSampler.sol

```diff
@@ -0,0 +1,99 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/DeploymentConstants.sol";
+import "./IDODO.sol";
+import "./SamplerUtils.sol";
+
+
+contract DODOSampler is
+    DeploymentConstants,
+    SamplerUtils
+{
+
+    /// @dev Gas limit for DODO calls.
+    uint256 constant private DODO_CALL_GAS = 150e3; // 150k
+
+    /// @dev Sample sell quotes from DODO.
+    /// @param takerToken Address of the taker token (what to sell).
+    /// @param makerToken Address of the maker token (what to buy).
+    /// @param takerTokenAmounts Taker token sell amount for each sample.
+    /// @return makerTokenAmounts Maker amounts bought at each taker token
+    ///         amount.
+    function sampleSellsFromDODO(
+        address takerToken,
+        address makerToken,
+        uint256[] memory takerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory makerTokenAmounts)
+    {
+        _assertValidPair(makerToken, takerToken);
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
+
+        // (Base token, quote token)
+        // DODO only supports selling base or buying base, i.e Sell ETH or Buy ETH
+        IDODO pool = IDODO(IDODOZoo(_getDODOZooAddress()).getDODO(takerToken, makerToken));
+        // If there is no pool then return early
+        if (address(pool) == address(0)) {
+            return makerTokenAmounts;
+        }
+
+        for (uint256 i = 0; i < numSamples; i++) {
+            (bool didSucceed, bytes memory resultData) =
+                address(pool).staticcall.gas(DODO_CALL_GAS)(
+                    abi.encodeWithSelector(
+                        IDODO(0).querySellBaseToken.selector,
+                        takerTokenAmounts[i]
+                    ));
+            uint256 buyAmount = 0;
+            if (didSucceed) {
+                buyAmount = abi.decode(resultData, (uint256));
+            }
+            // Exit early if the amount is too high for the source to serve
+            if (buyAmount == 0) {
+                break;
+            }
+            makerTokenAmounts[i] = buyAmount;
+        }
+    }
+
+    /// @dev Sample buy quotes from DODO.
+    /// @param takerToken Address of the taker token (what to sell).
+    /// @param makerToken Address of the maker token (what to buy).
+    /// @param makerTokenAmounts Maker token sell amount for each sample.
+    /// @return takerTokenAmounts Taker amounts sold at each maker token
+    ///         amount.
+    function sampleBuysFromDODO(
+        address takerToken,
+        address makerToken,
+        uint256[] memory makerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory takerTokenAmounts)
+    {
+        return takerTokenAmounts;
```

Comment:
Whilst buys (of ETH) are supported, we perform this as a sell of USDC. So not entirely useful.

TODO could support buy USDC by sellETH, but `buyAmount` is less frequently used